doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+ */
+SSL *SSL_EVENT_get_stream(SSL_EVENT *event);
+
+/*
+ * Works as normal if the SSL object has already had its BIOs configured.
+ * Otherwise it will check the new connection queue in the SSL_EVENT_CTX. If it
+ * finds a pending new connection then it will "accept" it, adding the BIOs and
+ * running the handshake. In the case of QUIC this will set up the SSL object as
+ * the "default" stream. Will return SSL_ERROR_WANT_READ in the event that no
+ * new conneciton is pending.
+ */
+__owur int SSL_accept(SSL *ssl);
+
+/*
+ * Works as normal for (D)TLS. For QUIC it is essentially the same but (as for
+ * SSL_accpet()) it will typically use the BIOs as set on the SSL_EVENT_CTX
+ * rather than having its own BIOs.
+ */
+__owur int SSL_connect(SSL *ssl);
+
+/*
+ * In the case of a (D)TLS client attempts a resumption handshake. Will always
+ * fail for a (D)TLS server (returns NULL). For QUIC it creates a new stream
+ * (works on either client or server).
+ * bidi - indicates whether the stream is bidirectional or write only. Selecting
+ *        write only is an error for (D)TLS
+ */
+SSL *SSL_new_stream(SSL *ssl, int bidi);
+
+/*
+ * Not supported for (D)TLS - will always return NULL. For QUIC it will check
+ * if a "new stream" event has occurred for the connection. If it has then a
+ * new SSL object is created to represent it and returned.
+ */
+SSL *SSL_accept_stream(SSL *ssl);
+
+````
+
+Proposed additions to the bio.h header file are as follows:
+
+````
+/*
+ * Send *nummsg messages (must be >0) from the local addresses to the peer
+ * addresses via the BIO b. local is an array of *nummsg local addresses to
+ * specify the local address each of the messages should be sent from. If local
+ * is NULL or an individual message local address is NULL then the default local
+ * address for the socket is used. Setting an explicit local address can be
+ * useful if the socket is bound with INADDR_ANY or IN6ADDR_ANY (i.e. multiple
+ * IP addresses are bound to the same socket). peer is an arry of *mummsg peer
+ * addresses to specify the peer address each of the messages should be sent to.
+ * If peer is NULL or an indivual message peer address is NULL then the BIO must
+ * be in a "connected" state. data is an array of *nummsg pointers. Each pointer
+ * is for the data for each individual message. Similarly dlen is array of
+ * *numsg lengths corresponding to each of the individual messages. After a
+ * successful call to this function *nummsg is updated with the number of
+ * messages that were successfully sent (which may be less than the total
+ * requested). It is thread-safe to call this function, or BIO_recvmmsg at the
+ * same time in multiple threads for the same BIO. It is not thread-safe to call
+ * this function at the same time as any other BIO function for the same BIO.
+ */
+int BIO_sendmmsg(BIO *b, void **data, size_t *dlen, size_t *nummsg,
+                 BIO_ADDR **peer, BIO_ADDR **local);
```

Comment:
@hlandau wrote:
> There has been some discussion recently of how to pass flags to send() via BIO, do we want a flags argument here?

*Maybe*. This is primarily going to be called from libssl. So how would libssl know what flags to use? The recent discussion had the flags to use be set via a separate ctrl. That might be a better solution because it enables the application to make the decision about any flags it might want without libssl having to know about them.